Add support for computing hashes using OpenSSL

jrohel opened the pull request:
It adds WITH_OPENSSL build option.
If it is ON, OpenSSL will be used instead of internal implementation
of computing hashes (MD5, SHA1, SHA224, SHA256, SHA384, SHA512).

jrohel committed:
Add support for computing hashes using OpenSSL

It adds WITH_OPENSSL build option.
If it is ON, OpenSSL will be used instead of internal implementation
of computing hashes (MD5, SHA1, SHA224, SHA256, SHA384, SHA512).

lukash commented on src/chksum.c:
```diff
@@ -15,10 +15,42 @@
+#define solv_MD5_Final(md, ctx) MD5_Final(md, ctx)
+#define solv_SHA1_Init(ctx) SHA1_Init(ctx)
+#define solv_SHA1_Update(ctx, data, len) SHA1_Update(ctx, data, len)
+#define solv_SHA1_Final(ctx, md) SHA1_Final(md, ctx)
```
Hmm, the arguments are flipped in this one... is it intentional?

jrohel replied:
Yes it is. Internal implementation of SHA1 has flipped arguments.

lukash replied:
Oh... clever :smile:

lukash commented on src/chksum.c:
```diff
@@ -15,10 +15,42 @@
+
+typedef SHA_CTX SHA1_CTX;
+typedef SHA256_CTX SHA224_CTX;
+typedef SHA512_CTX SHA384_CTX;
```
Probably a noob question, why these defines?

jrohel replied:
These types are defined in the internal implementation of hashes but not in OpenSSL. 
By defining these types, I can simple replace the internal implementation with OpenSSL (I have all needed types now). There is no need to additional code changes.

lukash replied:
I see, libsolv actually has the same typedefs. A short explanatory comment might be helpful here...

Otherwise, patch LGTM.

jrohel committed:
Use OpenSSL for computing hashes by default

It simplifies backporting to downstream, because
it does not need to change the .spec file for using OpenSSL.

Conan-Kudo reviewed (changes requested):
NAK. Upstream rejected this for adding bloat and complexity for simple functions that are trivial to validate. In principle, I agree with upstream on this, and we should not do this.

jrohel commented:
@Conan-Kudo 
This PR is for our downstream. It does not change API and behaviour of libsolv. -> The change is "invisible" for users of libsolv library. So there is not a problem with upstream compatibility. If you don't like it, you are not forced to use this patch in another distributions (projects).
Replacing a bundled code by library means adding bload? OK, I can simplify code. I can remove the WITH_OPENSSL option. OpenSSL will be allways used. Than I can remove 6 files (md5.c, md5.h, sha1.c, sha1.h, sha2.c, sha2.h) -> This files are not used if OpenSSL version is activated. So I will remove more then 1500 lines of bundled code from libsolv. Is that what you want?

> functions that are trivial to validate

Use validated library is better for security/certification than validate n private implementations.

pmatilai commented:
> NAK. Upstream rejected this for adding bloat and complexity for simple functions that are trivial to validate. In principle, I agree with upstream on this, and we should not do this.

This is absurd. Just because something is relatively simple doesn't mean it should be duplicated a thousand different projects, and calculating hashes isn't even *remotely* trivial in my books. 
We've  spent a lot of time and energy to eliminate these private implementations of cryptographic hashes so that we don't need to validate a thousand different things at great cost, or patch a thousand different things when a vulnerability is found. Just like we've eliminated all those bundled zlib implementations back in the day.

I can *sort of* understand libsolv upstream position as far as not wanting to add a hard dependency on openssl for a couple of hashes, but I think we should push upstream a bit harder to at least *allow* using it instead of private implementations. This is not like we're trying to pull in some 40GB extra library that wouldn't otherwise be there, openssl is *everywhere* already.

pmatilai commented:
Oh, by the way. Another possibility would be using the cryptographic API that rpm exports, when linked to librpm. That API is an abstraction that supports multiple backends, in current Fedora and RHEL that happens to be openssl.

Conan-Kudo commented:
@pmatilai I suspect that @mlschroe would be more amenable to leveraging librpm's API than OpenSSL's as an alternative to the bundled ones, since he already does require pulling librpm for various other reasons anyway.

Conan-Kudo commented:
@pmatilai At the same token, I'm pretty sure libsolv deliberately doesn't use _any_ crypto library so that it's maximally compatible with whatever people choose for cryptography when adding the downloader functions (i.e. if you're using libcurl, you could wind up with any given crypto library in your chain, and it could get ugly if you have a mixture of crypto libraries with matching symbols).

igor-raits commented:
I think one of main things is static linking. Any crypto libraries are pain to link statically. I think openssl is pretty bad in this regard (or used to be), esp. due to symbol conflicts (which had happened to me once while back in mesa, so I had to switch from openssl to gnutls and then I think they just implemented sha1 hashing in mesa directly).

Conan-Kudo commented:
@ignatenkobrain I forgot about that. Yes, that's a major problem since libsolv is statically linked in openSUSE for perl-BSSolv, libzypp, and libdnf.

(At least gnutls has properly namespaced symbols... OpenSSL wouldn't know symbol name/version discipline if it smacked them in the face...)

igor-raits commented:
```
This local copy of a SHA1 implementation based on the sources below.

Why:
 - Some libraries suffer from race condition and other issues. For example see
commit ade3108bb5b0 ("util: Fix race condition on libgcrypt initialization").

 - Fold the handling and detection of _eight_ implementations at configure
stage and _seven_ different codepaths.

 - Have a single, uniform, code used by developers, testers and users.

 - Avoid conflicts when using software which ships with it's own SHA1 library.
The latter of which conflicting with the one mesa is build against.
```

From mesa code.

dmach commented:
I don't understand what's the issue here.
The patch introduces opt-in linking with OpenSSL to avoid using custom hash functions.
We don't force anyone to change the defaults or to use it on other distros.
I understand that this wasn't accepted in upstream for a reason and we'll be maintaining this downstream patch here.
